src/merkle.rs

```diff
@@ -58,15 +58,66 @@ impl<'p, D: KeyValueDB, P: PoolParams> MerkleTree<'p, D, P> {
         I::IntoIter: 'a,
         P::Fr: 'a,
     {
-        let mut batch = self.db.transaction();
-
         for (index, hash, temporary) in hashes.into_iter().cloned() {
+            let mut batch = self.db.transaction();
             self.add_hash_batched(&mut batch, index, hash, temporary);
```

Comment:
I think `add_hash` can be used here instead. I'm not sure if `add_hash_batched` is needed anymore. Consider removing it in favor of the `add_hash` function.